Docs: Update plugin metadata (#24633)
grafana/grafana: docs/sources/developers/plugins/metadata.md

# plugin.json

> The plugin.json file is mandatory for all plugins. When Grafana starts it will scan the plugin folders and mount every folder that contains a plugin.json file unless the folder contains a subfolder named dist. In that case grafana will mount the dist folder instead.

Reviewer: ```suggestion
The plugin.json file is required for all plugins. When Grafana starts, it scans the plugin folders and mounts every folder that contains a plugin.json file unless the folder contains a subfolder named dist. In that case, Grafana mounts the dist folder instead.
```

## Properties

| Property        | Type                    | Required | Description                                                                                                                                                                                                                                                                       |
|-----------------|-------------------------|----------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `dependencies`  | [object](#dependencies) | **true** | Plugin dependencies.                                                                                                                                                                                                                                                              |

Reviewer 2: I guess one of the negative using go for generating this is that it cannot output properties in the order specified in the schema. If you generate again will there a different sort order?

Author: It sorts properties first by whether it's required or not, then alphabetically.

> | `id`            | string                  | **true** | Unique name of the plugin. If the plugin is published on Grafana.com then the plugin id has to follow the naming conventions.                                                                                                                                                     |

Reviewer: ```suggestion
| `id`            | string                  | **true** | Unique name of the plugin. If the plugin is published on Grafana.com, then the plugin id has to follow the naming conventions.                                                                                                                                                     |
```

> | `info`          | [object](#info)         | **true** | Meta data for the plugin. Some fields are used on the plugins page in Grafana and others on Grafana.com if the plugin is published.                                                                                                                                               |

Reviewer: ```suggestion
| `info`          | [object](#info)         | **true** | Metadata for the plugin. Some fields are used on the plugins page in Grafana and others on Grafana.com if the plugin is published.                                                                                                                                               |
```

| `name`          | string                  | **true** | Human-readable name of the plugin that is shown to the user in the UI.                                                                                                                                                                                                            |

> | `type`          | string                  | **true** | Plugin type. Possible values are `app`, `datasource`, `panel`.                                                                                                                                                                                                                    |

Reviewer: ```suggestion
| `type`          | string                  | **true** | Plugin type. Possible values are `app`, `datasource`, and `panel`.                                                                                                                                                                                                                    |
```

| `alerting`      | boolean                 | false    | For data source plugins. If the plugin supports alerting.                                                                                                                                                                                                                         |
| `annotations`   | boolean                 | false    | For data source plugins. If the plugin supports annotation queries.                                                                                                                                                                                                               |
| `backend`       | boolean                 | false    | If the plugin has a backend component.                                                                                                                                                                                                                                            |

> | `category`      | string                  | false    | Plugin category used on the Add data source page. Possible values are: `tsdb`, `logging`, `cloud`, `tracing`, `sql`.                                                                                                                                                              |

Reviewer: ```suggestion
| `category`      | string                  | false    | Plugin category used on the Add data source page. Possible values are: `tsdb`, `logging`, `cloud`, `tracing`, and `sql`.                                                                                                                                                              |
```

> | `executable`    | string                  | false    | The first part of the file name of the backend component executables (there can three binaries for linux, darwin and windows).                                                                                                                                                    |

Reviewer: ```suggestion
| `executable`    | string                  | false    | The first part of the file name of the backend component executables (there can three binaries for Linux, Darwin, and Windows).                                                                                                                                                    |
```

| `tables`        | boolean                 | false    |                                                                                                                                                                                                                                                                                   |
| `tracing`       | boolean                 | false    | For data source plugins. If the plugin supports streaming.                                                                                                                                                                                                                        |

## dependencies

Plugin dependencies.

### Properties

| Property         | Type   | Required | Description                                                |
|------------------|--------|----------|------------------------------------------------------------|

> | `grafanaVersion` | string | **true** | Required Grafana version for this plugin, e.g. `6.7.x`     |

Reviewer: ```suggestion
| `grafanaVersion` | string | **true** | Required Grafana version for this plugin, e.g. `6.7.x`.     |
```

| `plugins`        | array  | false    | An array of required plugins on which this plugin depends. |

## includes

### Properties

| Property     | Type    | Required | Description |
|--------------|---------|----------|-------------|
| `addToNav`   | boolean | false    |             |
| `component`  | string  | false    |             |
| `defaultNav` | boolean | false    |             |
| `name`       | string  | false    |             |
| `role`       | string  | false    |             |
| `type`       | string  | false    |             |

> ## info

Reviewer: ```suggestion
## Info
```

Author: Due to a limitation in the generator tool, this is not possible for now. I'll look into improving it later on.

> Meta data for the plugin. Some fields are used on the plugins page in Grafana and others on Grafana.com if the plugin is published.

Reviewer: ```suggestion
Metadata for the plugin. Some fields are used on the plugins page in Grafana and others on Grafana.com if the plugin is published.
```

### Properties

| Property      | Type              | Required | Description                                                                                                                   |
|---------------|-------------------|----------|-------------------------------------------------------------------------------------------------------------------------------|
| `logos`       | [object](#logos)  | **true** | SVG images that are used as plugin icons.                                                                                     |
| `updated`     | string            | **true** | Date when this plugin was built.                                                                                              |

> | `version`     | string            | **true** | Project version of this commit, e.g. `6.7.x`                                                                                  |

Reviewer: ```suggestion
| `version`     | string            | **true** | Project version of this commit, e.g. `6.7.x`.                                                                              |
```

| `author`      | [object](#author) | false    | Information about the plugin author.                                                                                          |

> | `description` | string            | false    | Description of plugin. Used on the plugins page in Grafana and for search on grafana.com.                                     |

Reviewer: Here "grafana.com" is lowercase. Elsewhere, "Grafana.com" is uppercase. Reconcile throughout.

| `minInterval`   | boolean | false    | For data source plugins. If the `min interval` option should be shown in the query options section in the query editor.    |

## routes

For data source plugins. Proxy routes used for plugin authentication and adding headers to HTTP requests made by the plugin. For more information, refer to [Authentication for data source plugins](https://grafana.com/docs/grafana/latest/developers/plugins/authentication/).

### Properties

| Property      | Type                 | Required | Description                                                                                             |
|---------------|----------------------|----------|---------------------------------------------------------------------------------------------------------|
| `headers`     | array                | false    | For data source plugins. Route headers adds HTTP headers to the proxied request.                        |
| `method`      | string               | false    | For data source plugins. Route method matches the HTTP verb like GET or POST.                           |

> | `path`        | string               | false    | For data source plugins. The route path that is replaced by the route url field when proxying the call. |

Reviewer: ```suggestion
| `path`        | string               | false    | For data source plugins. The route path that is replaced by the route URL field when proxying the call. |
```

| `reqRole`     | string               | false    |                                                                                                         |
| `reqSignedIn` | boolean              | false    |                                                                                                         |
| `tokenAuth`   | [object](#tokenauth) | false    | For data source plugins. Token authentication section used with an OAuth API.                           |

> | `url`         | string               | false    | For data source plugins. Route url is where the request is proxied to.                                  |

Reviewer: ```suggestion
| `url`         | string               | false    | For data source plugins. Route URL is where the request is proxied to.                                  |
```

### tokenAuth

For data source plugins. Token authentication section used with an OAuth API.

#### Properties

| Property | Type              | Required | Description                                                           |
|----------|-------------------|----------|-----------------------------------------------------------------------|
| `params` | [object](#params) | false    | For data source plugins. Params for the token authentication request. |

> | `url`    | string            | false    | For data source plugins. Url to fetch the authentication token.       |

Reviewer: ```suggestion
| `url`    | string            | false    | For data source plugins. URL to fetch the authentication token.       |
```

#### params

> For data source plugins. Params for the token authentication request.

Reviewer: Write out "parameters" everywhere that it is just shortening the word rather than referring to something in the code.

| `type`         | string                  | **true** | Plugin type. Possible values are `app`, `datasource`, `panel`.                                                                       |
| `alerting`     | boolean                 | false    | For data source plugins. If the plugin supports alerting.                                                                            |
| `annotations`  | boolean                 | false    | For data source plugins. If the plugin supports annotation queries.                                                                  |
| `backend`      | boolean                 | false    | If the plugin has a backend component.                                                                                               |
| `category`     | string                  | false    | Plugin category used on the Add data source page. Possible values are: `tsdb`, `logging`, `cloud`, `tracing`, `sql`.                 |
| `executable`   | string                  | false    | The first part of the file name of the backend component executables (there can three binaries for linux, darwin and windows).       |

> | `hasQueryHelp` | boolean                 | false    | For data source plugins. TODO not sure if this is used.                                                                              |

Reviewer 3: Can be deleted is not used anywhere

Author: Should we mark these as deprecated. I ran the schema against all the community plugins, and there's additional fields like `partials`, and `includes`. Maybe we should include them, but with a deprecated flag?

Reviewer 2: Includes is used for embedding dashboards with plugins, right? Partials not sure about, what plugin have that?

Author: There are 35 plugins using `partials`, including the [simple-json-datasource](https://github.com/grafana/simple-json-datasource/blob/master/src/plugin.json#L6-L8).

There are 15 plugins using `includes`, including the [kubernetes-app](https://github.com/grafana/kubernetes-app/blob/master/src/plugin.json#L37-L80).

Author: Interestingly enough, out of 91 community plugins, only 60 plugins conform this schema.

Majority of the ones that don't conform have the following issues:

- Many are missing `dependencies` or `logos`
- Examples of invalid versions include "^0.0.7", ">=6.3.0", "4.1.1+", "5.0+"
- `updated` has examples of both "2001-01-01", and "01-01-01"

| `routes`       | array                   | false    | For data source plugins. Proxy routes used for plugin authentication and adding headers to HTTP requests made by the plugin.         |
| `sort`         | number                  | false    | Internal property for sorting. Cannot be used as will be overwritten by Grafana.                                                     |
| `streaming`    | boolean                 | false    | For data source plugins. If the plugin supports streaming.                                                                           |
| `tracing`      | boolean                 | false    | For data source plugins. If the plugin supports streaming.                                                                           |

## dependencies

Plugin dependencies.

### Properties

| Property         | Type   | Required | Description                                                |
|------------------|--------|----------|------------------------------------------------------------|

> | `grafanaVersion` | string | **true** | Required Grafana version for this plugin. TODO - format.   |

Reviewer 3: I think @marefr can help here. Should be something something semver :-)

Reviewer 2: The grafana.com have some logic for showing Grafana dependency and seems it works best if you use a format of `major.minor.path`, for example:
`Required Grafana version for this plugin. For example `6.x.x 7.x.x` if plugin requires Grafana v6.x.x and v7.x.x.`

For reference see Grafana dependency in right column for https://grafana.com/grafana/plugins/raintank-worldping-app

Author: Added a regex validator to versions, as well as examples in the description.

| `small`  | string | **true** | Link to the "small" version of the plugin logo, which must be an SVG image. "Large" and "small" logos can be the same image. |

## queryOptions

For data source plugins. There is a query options section in the plugin's query editor and these options can be turned on if needed.

### Properties

| Property        | Type    | Required | Description                                                                                                                |
|-----------------|---------|----------|----------------------------------------------------------------------------------------------------------------------------|
| `cacheTimeout`  | boolean | false    | For data source plugins. If the `cache timeout` option should be shown in the query options section in the query editor.   |
| `maxDataPoints` | boolean | false    | For data source plugins. If the `max data points` option should be shown in the query options section in the query editor. |
| `minInterval`   | boolean | false    | For data source plugins. If the `min interval` option should be shown in the query options section in the query editor.    |

## routes

> For data source plugins. Proxy routes used for plugin authentication and adding headers to HTTP requests made by the plugin.

Reviewer 3: Should link to https://grafana.com/docs/grafana/latest/developers/plugins/authentication/
